File: backend/src/services/embedding-service.ts
import type { PoolClient } from "pg";
import { embedBatchCached } from "../ai/embedder";
import { getEmbeddingProvider } from "../ai";
import { createLogger } from "../utils/logger";

const log = createLogger("embedding-service");

/**
 * EmbeddingService manages on-demand embedding of indexed documents.
 * Handles:
 * - Fetching unembed chunks from already-indexed documents
 * - Generating embeddings via OpenAI
 * - Persisting embeddings to database
 */

interface DocumentChunk {
  id: string;
  content: string;
  embedding: unknown;
  chunk_index: number;
}

interface EmbedDocumentResult {
  documentId: string;
  totalChunks: number;
  embeddedChunks: number;
  skippedChunks: number;
  status: "success" | "no-chunks" | "embeddings-disabled" | "failed";
}

/**
 * Embed all chunks for a single document.
 * Skips chunks that already have embeddings.
 * Returns status with counts.
 */
export async function embedDocument(
  client: PoolClient,
  documentId: string,
  orgId: string
): Promise<EmbedDocumentResult> {
  const provider = getEmbeddingProvider();
  if (!provider) {
    log.warn({ documentId }, "Embeddings disabled, skipping");
    return {
      documentId,
      totalChunks: 0,
      embeddedChunks: 0,
      skippedChunks: 0,
      status: "embeddings-disabled",
    };
  }

  // Fetch all chunks for this document
  const chunksResult = await client.query<DocumentChunk>(
    `SELECT id, content, embedding, chunk_index
     FROM document_chunks
     WHERE document_id = $1 AND org_id = $2
     ORDER BY chunk_index ASC`,
    [documentId, orgId]
  );

  const chunks = chunksResult.rows;
  if (chunks.length === 0) {
    log.info({ documentId }, "No chunks found for document");
    return {
      documentId,
      totalChunks: 0,
      embeddedChunks: 0,
      skippedChunks: 0,
      status: "no-chunks",
    };
  }

  // Separate chunks that need embedding
  const unembed: { id: string; content: string; index: number }[] = [];
  for (const chunk of chunks) {
    if (!chunk.embedding) {
      unembed.push({
        id: chunk.id,
        content: chunk.content,
        index: chunk.chunk_index,
      });
    }
  }

  if (unembed.length === 0) {
    log.info({ documentId, totalChunks: chunks.length }, "All chunks already embedded");
    return {
      documentId,
      totalChunks: chunks.length,
      embeddedChunks: 0,
      skippedChunks: chunks.length,
      status: "success",
    };
  }

  log.info(
    { documentId, totalChunks: chunks.length, unembedded: unembed.length },
    "Starting embedding batch"
  );

  // Generate embeddings for unembed chunks
  const texts = unembed.map((c) => c.content);
  const embeddings = await embedBatchCached(texts, client);

  // Update database with embeddings
  for (let i = 0; i < unembed.length; i += 1) {
    const chunk = unembed[i];
    const embedding = embeddings[i];

    await client.query(
      `UPDATE document_chunks
       SET embedding = $1
       WHERE id = $2`,
      [embedding ? JSON.stringify(embedding) : null, chunk.id]
    );
  }

  log.info(
    { documentId, totalChunks: chunks.length, embedded: unembed.length },
    "Document embedding complete"
  );

  return {
    documentId,
    totalChunks: chunks.length,
    embeddedChunks: unembed.length,
    skippedChunks: chunks.length - unembed.length,
    status: "success",
  };
}

/**
 * Embed all documents in an organization (batch operation).
 * Useful for backfilling embeddings when feature is enabled.
 */
export async function embedOrganization(
  client: PoolClient,
  orgId: string
): Promise<{
  totalDocuments: number;
  embeddedDocuments: number;
  failedDocuments: number;
  totalChunks: number;
  embeddings: EmbedDocumentResult[];
}> {
  log.info({ orgId }, "Starting organization-wide embedding");

  // Fetch all documents in org
  const docsResult = await client.query<{ id: string }>(
    `SELECT id FROM documents WHERE org_id = $1 ORDER BY indexed_at DESC`,
    [orgId]
  );

  const docIds = docsResult.rows.map((r) => r.id);
  const results: EmbedDocumentResult[] = [];
  let totalChunks = 0;
  let failedCount = 0;

  for (const docId of docIds) {
    try {
      const result = await embedDocument(client, docId, orgId);
      results.push(result);
      totalChunks += result.totalChunks;
    } catch (error) {
      log.error({ docId, error }, "Failed to embed document");
      failedCount += 1;
      results.push({
        documentId: docId,
        totalChunks: 0,
        embeddedChunks: 0,
        skippedChunks: 0,
        status: "failed",
      });
    }
  }

  const embeddedCount = results.filter((r) => r.embeddedChunks > 0).length;

  log.info(
    {
      orgId,
      totalDocuments: docIds.length,
      embedded: embeddedCount,
      failed: failedCount,
      totalChunks,
    },
    "Organization embedding complete"
  );

  return {
    totalDocuments: docIds.length,
    embeddedDocuments: embeddedCount,
    failedDocuments: failedCount,
    totalChunks,
    embeddings: results,
  };
}

export async function embedConnector(
  client: PoolClient,
  orgId: string,
  connectorId: string
): Promise<{
  totalDocuments: number;
  embeddedDocuments: number;
  failedDocuments: number;
  totalChunks: number;
  embeddings: EmbedDocumentResult[];
}> {
  log.info({ orgId, connectorId }, "Starting connector embedding");

  const docsResult = await client.query<{ id: string }>(
    `SELECT id
     FROM documents
     WHERE org_id = $1 AND connector_id = $2
     ORDER BY indexed_at DESC`,
    [orgId, connectorId]
  );

  const docIds = docsResult.rows.map((r) => r.id);
  const results: EmbedDocumentResult[] = [];
  let totalChunks = 0;
  let failedCount = 0;

  for (const docId of docIds) {
    try {
      const result = await embedDocument(client, docId, orgId);
      results.push(result);
      totalChunks += result.totalChunks;
    } catch (error) {
      log.error({ docId, connectorId, error }, "Failed to embed document");
      failedCount += 1;
      results.push({
        documentId: docId,
        totalChunks: 0,
        embeddedChunks: 0,
        skippedChunks: 0,
        status: "failed",
      });
    }
  }

  const embeddedCount = results.filter((r) => r.embeddedChunks > 0).length;

  return {
    totalDocuments: docIds.length,
    embeddedDocuments: embeddedCount,
    failedDocuments: failedCount,
    totalChunks,
    embeddings: results,
  };
}

File: backend/app/services/embedding_service.py
from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Any, Literal

from sqlalchemy import select, update

from app.ai.embedder import embed_batch_cached
from app.ai.index import get_embedding_provider
from app.db.tables import document_chunks, documents

log = logging.getLogger("embedding-service")


@dataclass
class EmbedDocumentResult:
    document_id: str
    total_chunks: int
    embedded_chunks: int
    skipped_chunks: int
    status: Literal["success", "no-chunks", "embeddings-disabled", "failed"]


async def embed_document(
    conn,
    document_id: str,
    org_id: str,
) -> EmbedDocumentResult:
    provider = get_embedding_provider()
    if provider is None:
        log.warning("Embeddings disabled, skipping document_id=%s", document_id)
        return EmbedDocumentResult(
            document_id=document_id,
            total_chunks=0,
            embedded_chunks=0,
            skipped_chunks=0,
            status="embeddings-disabled",
        )

    stmt = (
        select(
            document_chunks.c.id,
            document_chunks.c.content,
            document_chunks.c.embedding,
            document_chunks.c.chunk_index,
        )
        .where(
            document_chunks.c.document_id == document_id,
            document_chunks.c.org_id == org_id,
        )
        .order_by(document_chunks.c.chunk_index.asc())
    )
    chunks = await conn.fetch(stmt)

    if not chunks:
        log.info("No chunks found for document_id=%s", document_id)
        return EmbedDocumentResult(
            document_id=document_id,
            total_chunks=0,
            embedded_chunks=0,
            skipped_chunks=0,
            status="no-chunks",
        )

    unembed = [
        {"id": str(row["id"]), "content": row["content"], "index": row["chunk_index"]}
        for row in chunks
        if not row["embedding"]
    ]

    if not unembed:
        log.info("All chunks already embedded document_id=%s total=%d", document_id, len(chunks))
        return EmbedDocumentResult(
            document_id=document_id,
            total_chunks=len(chunks),
            embedded_chunks=0,
            skipped_chunks=len(chunks),
            status="success",
        )

    log.info(
        "Starting embedding batch document_id=%s total=%d unembedded=%d",
        document_id,
        len(chunks),
        len(unembed),
    )

    texts = [c["content"] for c in unembed]
    embeddings = await embed_batch_cached(texts, conn)

    for i, chunk in enumerate(unembed):
        embedding = embeddings[i] if i < len(embeddings) else []
        # Pass the Python list directly — SQLAlchemy handles JSONB serialisation.
        # Use None when there is no embedding so the column stays NULL.
        embedding_value = embedding if embedding else None
        await conn.execute(
            update(document_chunks)
            .where(document_chunks.c.id == chunk["id"])
            .values(embedding=embedding_value)
        )

    log.info(
        "Document embedding complete document_id=%s total=%d embedded=%d",
        document_id,
        len(chunks),
        len(unembed),
    )

    return EmbedDocumentResult(
        document_id=document_id,
        total_chunks=len(chunks),
        embedded_chunks=len(unembed),
        skipped_chunks=len(chunks) - len(unembed),
        status="success",
    )


async def embed_organization(
    conn,
    org_id: str,
) -> dict[str, Any]:
    log.info("Starting organization-wide embedding org_id=%s", org_id)

    doc_rows = await conn.fetch(
        select(documents.c.id)
        .where(documents.c.org_id == org_id)
        .order_by(documents.c.indexed_at.desc())
    )

    results: list[EmbedDocumentResult] = []
    total_chunks = 0
    failed_count = 0

    for row in doc_rows:
        doc_id = str(row["id"])
        try:
            result = await embed_document(conn, doc_id, org_id)
            results.append(result)
            total_chunks += result.total_chunks
        except Exception as exc:
            log.error("Failed to embed document doc_id=%s error=%s", doc_id, exc)
            failed_count += 1
            results.append(
                EmbedDocumentResult(
                    document_id=doc_id,
                    total_chunks=0,
                    embedded_chunks=0,
                    skipped_chunks=0,
                    status="failed",
                )
            )

    embedded_count = sum(1 for r in results if r.embedded_chunks > 0)
    return {
        "total_documents": len(doc_rows),
        "embedded_documents": embedded_count,
        "failed_documents": failed_count,
        "total_chunks": total_chunks,
        "embeddings": [vars(r) for r in results],
    }


async def embed_connector(
    conn,
    org_id: str,
    connector_id: str,
) -> dict[str, Any]:
    log.info("Starting connector embedding org_id=%s connector_id=%s", org_id, connector_id)

    doc_rows = await conn.fetch(
        select(documents.c.id)
        .where(documents.c.org_id == org_id, documents.c.connector_id == connector_id)
        .order_by(documents.c.indexed_at.desc())
    )

    results: list[EmbedDocumentResult] = []
    total_chunks = 0
    failed_count = 0

    for row in doc_rows:
        doc_id = str(row["id"])
        try:
            result = await embed_document(conn, doc_id, org_id)
            results.append(result)
            total_chunks += result.total_chunks
        except Exception as exc:
            log.error(
                "Failed to embed document doc_id=%s connector_id=%s error=%s",
                doc_id,
                connector_id,
                exc,
            )
            failed_count += 1
            results.append(
                EmbedDocumentResult(
                    document_id=doc_id,
                    total_chunks=0,
                    embedded_chunks=0,
                    skipped_chunks=0,
                    status="failed",
                )
            )

    embedded_count = sum(1 for r in results if r.embedded_chunks > 0)
    return {
        "total_documents": len(doc_rows),
        "embedded_documents": embedded_count,
        "failed_documents": failed_count,
        "total_chunks": total_chunks,
        "embeddings": [vars(r) for r in results],
    }
